source/libuv/uv_event_loop.c

```diff
@@ -0,0 +1,792 @@
+}
+
+static int s_running_tasks_start(void *context, struct aws_hash_element *element) {
+
+    struct task_data *task = element->value;
+    struct libuv_loop *impl = context;
+
+    uv_timer_start(
+        &task->timer, s_uv_task_timer_cb, s_timestamp_to_uv_millis(task->event_loop, task->task->timestamp), 0);
+
+    aws_atomic_fetch_add(&impl->num_open_handles, 1);
+
+    return AWS_COMMON_HASH_TABLE_ITER_CONTINUE;
+}
+
+static int s_run(struct aws_event_loop *event_loop) {
+
+    bool cleanup_polls = false;
+
+    struct libuv_loop *impl = event_loop->impl_data;
+
+    impl->schedule_tasks_async.data = impl;
+    if (uv_async_init(impl->uv_loop, &impl->schedule_tasks_async, s_uv_async_schedule_tasks)) {
+        return AWS_OP_ERR;
+    }
+    aws_atomic_fetch_add(&impl->num_open_handles, 1);
+
+    /* Prep the stop async */
+    impl->stop_async.data = impl;
+    if (uv_async_init(impl->uv_loop, &impl->stop_async, s_uv_async_stop_loop)) {
+        goto clean_up;
+    }
+    aws_atomic_fetch_add(&impl->num_open_handles, 1);
+
+    /* Start all existing subscriptions */
+    struct aws_linked_list_node *open_subs_it = aws_linked_list_begin(&impl->el_thread_data.open_subscriptions);
+    const struct aws_linked_list_node *open_subs_end = aws_linked_list_end(&impl->el_thread_data.open_subscriptions);
+    while (open_subs_it != open_subs_end) {
+        struct handle_data *handle_data = AWS_CONTAINER_OF(open_subs_it, struct handle_data, open_subs_node);
+        if (uv_poll_start(&handle_data->poll, handle_data->uv_events, s_uv_poll_cb)) {
+            goto clean_up;
+        }
+        aws_atomic_fetch_add(&impl->num_open_handles, 1);
+
+        open_subs_it = open_subs_it->next;
+    }
+    cleanup_polls = true;
+
+    /* Start all existing timers */
+    aws_hash_table_foreach(&impl->el_thread_data.running_tasks, s_running_tasks_start, impl);
+
+    if (impl->owns_uv_loop) {
+        assert(s_owned(impl)->state == EVENT_THREAD_STATE_READY_TO_RUN);
+
+        if (aws_thread_launch(&s_owned(impl)->thread, &s_thread_loop, impl, NULL)) {
+            goto clean_up;
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+
+clean_up:
+
+    if (cleanup_polls) {
+        struct aws_linked_list_node *open_subs_it = aws_linked_list_begin(&impl->el_thread_data.open_subscriptions);
+        const struct aws_linked_list_node *open_subs_end =
+            aws_linked_list_end(&impl->el_thread_data.open_subscriptions);
+        while (open_subs_it != open_subs_end) {
+            struct handle_data *handle_data = AWS_CONTAINER_OF(open_subs_it, struct handle_data, open_subs_node);
+
+            uv_close((uv_handle_t *)&handle_data->poll, s_uv_close_sub);
+
+            open_subs_it = open_subs_it->next;
+        }
+    }
+    if (impl->stop_async.loop) {
+        uv_close((uv_handle_t *)&impl->stop_async, s_uv_close_handle);
+    }
+    uv_close((uv_handle_t *)&impl->cross_thread_data, s_uv_close_handle);
```

Comment:
cross_thread_data is a uv_handle_t? huh?

Reply:
FUCK C CASTS

Reply:
🥇

Reply:
I saw this last night, and I assumed it was the first member of the struct so it was fine, I'm a bad person for thinking that was ok :(